File: ZezeJava/ZezeJava/src/main/java/Zeze/Component/RedoQueueServer.java
package Zeze.Component;

import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Predicate;
import Zeze.Application;
import Zeze.Net.Binary;
import Zeze.Services.HandshakeServer;
import Zeze.Transaction.Procedure;
import Zeze.Util.LongConcurrentHashMap;

public class RedoQueueServer extends AbstractRedoQueueServer {
	private final ConcurrentHashMap<String, LongConcurrentHashMap<Predicate<Binary>>> handles = new ConcurrentHashMap<>();
	private final Server server;

	public RedoQueueServer(Application zeze) {
		server = new Server(zeze);
		RegisterProtocols(server);
		RegisterZezeTables(zeze);
	}

	@Override
	public void UnRegister() {
		UnRegisterProtocols(server);
		UnRegisterZezeTables(server.getZeze());
	}

	public void start() throws Exception {
		server.start();
	}

	public void stop() throws Exception {
		server.stop();
	}

	/**
	 * 注册任务，
	 */
	public void register(String queue, int type, Predicate<Binary> task) {
		if (null != handles.computeIfAbsent(queue, __ -> new LongConcurrentHashMap<>()).putIfAbsent(type, task))
			throw new IllegalStateException("duplicate task type. " + type);
	}

	@Override
	protected long ProcessRunTaskRequest(Zeze.Builtin.RedoQueue.RunTask r) {
		var last = _tQueueLastTaskId.getOrAdd(r.Argument.getQueueName());
		r.Result.setTaskId(last.getTaskId());
		if (r.Argument.getPrevTaskId() != last.getTaskId())
			return Procedure.ErrorRequestId;
		var queue = handles.get(r.Argument.getQueueName());
		if (queue == null)
			return Procedure.NotImplement;
		var handle = queue.get(r.Argument.getTaskType());
		if (handle == null)
			return Procedure.NotImplement;
		if (!handle.test(r.Argument.getTaskParam()))
			return Procedure.LogicError;
		last.setTaskId(r.Argument.getTaskId());
		r.Result.setTaskId(last.getTaskId());
		return Procedure.Success;
	}

	public static class Server extends HandshakeServer {
		public Server(Application zeze) {
			super("RedoQueueServer", zeze);
		}
		// 不覆写dispatchProtocol（CP1-F2）：基类对事务级协议copy网络buffer后
		// 在procedure内重解码（Service.dispatchProtocol(long,ByteBuffer,...)），redo安全且
		// 不异步引用可回收的网络缓冲；RunTask的factoryHandle.Level默认Serializable，
		// 事务语义与原覆写一致。原覆写在procedure内直接设bb.ReadIndex=0重解码网络buffer，
		// 派发入池后缓冲可能被回收复用，解码出错误协议内容。
	}
}

File: Zeze/Component/RedoQueueServer.cs
using System;
using System.Collections.Concurrent;
using System.Threading.Tasks;
using Zeze.Builtin.RedoQueue;
using Zeze.Net;
using Zeze.Util;

namespace Zeze.Component
{
	public class RedoQueueServer : AbstractRedoQueueServer
	{
		protected override async Task<long> ProcessRunTaskRequest(Protocol _p)
		{
			var r = _p as RunTask;
			Transaction.Transaction.Current.RunWhileCommit(() => r.SendResult());

			var last = await _tQueueLastTaskId.GetOrAddAsync(r.Argument.QueueName);
			r.Result.TaskId = last.TaskId;
			if (r.Argument.PrevTaskId != last.TaskId)
				return ResultCode.ErrorRequestId;
			if (!handles.TryGetValue(r.Argument.QueueName, out var queue))
				return ResultCode.NotImplement;
			if (!queue.TryGetValue(r.Argument.TaskType, out var handle))
				return ResultCode.NotImplement;
			if (!handle(r.Argument.TaskParam))
				return ResultCode.LogicError;
			last.TaskId = r.Argument.TaskId;
			r.Result.TaskId = last.TaskId;
			return ResultCode.Success;
		}

		private readonly ConcurrentDictionary<string, ConcurrentDictionary<int, Func<Binary, bool>>> handles = new();
		private Server server;

		public RedoQueueServer(Application zeze)
		{
			server = new Server(zeze);
			RegisterProtocols(server);
			RegisterZezeTables(zeze);
		}

		public override void UnRegister()
		{
			UnRegisterProtocols(server);
			UnRegisterZezeTables(server.Zeze);
		}

		public void Start()
		{
			server.Start();
		}

		public void Stop()
		{
			server.Stop();
		}

		/**
		 * ×¢²áÈÎÎñ£¬
		 */
		public void Register(string queue, int type, Func<Binary, bool> task)
		{
			if (!handles.GetOrAdd(queue, _ => new()).TryAdd(type, task))
				throw new Exception("duplicate task type. " + type);
		}

		public class Server : Services.HandshakeServer
		{
			public Server(Application zeze)
				: base("RedoQueueServer", zeze)
			{
			}

			public override void DispatchProtocol(Protocol p, ProtocolFactoryHandle factoryHandle)
			{
				var r = p as RunTask;
				var proc = Zeze.NewProcedure(async () => await factoryHandle.Handle(p),
					$"RedoQueue={r.Argument.QueueName} RunTask={r.Argument.TaskType}");
				_ = Mission.CallAsync(proc, p, (p, code) => p.TrySendResultCode(code)); // error result
			}
		}
	}
}
